File: 1022-sum-of-root-to-leaf-binary-numbers/1022-sum-of-root-to-leaf-binary-numbers.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    void rootToLeaf(TreeNode* root, string s, int* ans){
        if(!root->left && !root->right){
            s += to_string(root->val);
            ans[0]+=stoi(s,0,2);
            return;
        }
        string curr = to_string(root->val);
        if(root->left) rootToLeaf(root->left,s+curr,ans);
        if(root->right) rootToLeaf(root->right,s+curr,ans);
    }
    
    int sumRootToLeaf(TreeNode* root) {
        int* ans = new int[1];
        ans[0] = 0;
        rootToLeaf(root,"",ans);
        return ans[0];
    }
};

//--------------Another approach-------------------------------------------------------------
// class Solution {
//     int sumRootToLeaf(TreeNode* root, int sum){
//         if(!root) return 0;
//         sum = (2 * sum) + root->val;
//         if(!root->left && !root->right) return sum;
//         return sumRootToLeaf(root->left, sum) + sumRootToLeaf(root->right, sum);
//     }
// public:
//     int sumRootToLeaf(TreeNode* root) {
//         return sumRootToLeaf(root, 0);
//     }
    
// };

File: 1022-sum-of-root-to-leaf-binary-numbers/1022-sum-of-root-to-leaf-binary-numbers.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def solve(self,root, res):
        if root == None: return 0
        res = (2 * res) + root.val
        if root.left == None and root.right == None: return res
        return self.solve(root.left, res) + self.solve(root.right, res)
    
    def sumRootToLeaf(self, root: Optional[TreeNode]) -> int:
        return self.solve(root, 0)